soql-standalone-parser/src/main/scala/com/socrata/soql/BinaryTree.scala

```diff
@@ -0,0 +1,90 @@
+package com.socrata.soql
+
+trait BinaryTree[+T] {
+
+  def seq: Seq[T] = {
+    Seq(asLeaf)
+  }
+
+  /**
+   * output schema
+   */
+  def previous: T = {
+    asLeaf
+  }
+
+  def last: T = asLeaf
+
+  def asLeaf: T = {
+    this match {
+      case _: Compound[T] => ???
+      case t: T @unchecked => t
+    }
+  }
```

Comment:
Ugh no.  Partial functions are bad, this should return an `Option[T]` so if it's not a leaf use-sites are forced to deal with that possibilty.

Reply:
(or better still not exist at all and just require use-sites to pattern-match out the leaves)